supervision/dataset/core.py

```diff
@@ -80,21 +134,30 @@ def __iter__(self) -> Iterator[Tuple[str, np.ndarray, Detections]]:
                 An iterator that yields tuples containing the image name,
                 the image data, and its corresponding annotation.
         """
-        for image_name, image in self.images.items():
-            yield image_name, image, self.annotations.get(image_name, None)
+        for i in range(len(self)):
+            image_path, image, annotation = self[i]
+            yield image_path, image, annotation
 
-    def __eq__(self, other):
+    def __eq__(self, other) -> bool:
         if not isinstance(other, DetectionDataset):
             return False
 
         if set(self.classes) != set(other.classes):
             return False
 
-        for key in self.images:
-            if not np.array_equal(self.images[key], other.images[key]):
-                return False
-            if not self.annotations[key] == other.annotations[key]:
-                return False
+        if self.image_paths != other.image_paths:
+            return False
+
+        try:
+            for self_values, other_values in zip(self, other):
```

Comment:
This loads images. I think we should adjust this to compare images if already loaded.